rustls-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,168 @@
+//! EcDh functions. For further documentation please refer to rust_symcrypt::ecdh
+use rustls::crypto::{ActiveKeyExchange, GetRandomFailed, SharedSecret, SupportedKxGroup};
+use rustls::{Error, NamedGroup};
+
+use symcrypt::ecdh::EcDh;
+use symcrypt::eckey::CurveType;
+
+/// KxGroup is a struct that easily ties `rustls::NamedGroup` to the `symcrypt_sys::ecurve::CurveType`.
+///
+/// This outlines the supported key exchange groups that are exposed by Rustls and implemented by SymCrypt.
+/// Currently the only supported key exchange groups are `X25519`, `NistP38`, and `Nist256`.
+#[derive(Debug)]
+pub struct KxGroup {
+    name: NamedGroup,
+    curve_type: CurveType,
+}
+
+/// KeyExchange is a struct that defines the state for EcDh operations
+///
+/// the `state` field is tied to the [`EcDh`] struct from symcrypt_sys.
+///
+/// the `name` and `curve_type` provide access to the `rustls::NamedGroup`
+/// and `symcrypt_sys::ecurve::CurveType` respectively
+///
+/// `pub_key` is a `Vec<u8>` that represents the public key that is tied to the [`EcDh`] state.
+/// The `private_key` is not exposed.
+pub struct KeyExchange {
+    state: EcDh,
+    name: NamedGroup,
+    curve_type: CurveType,
+    pub_key: Vec<u8>,
+}
+
+/// All supported KeyExchange groups.
+/// ```ignore
+/// SECP384R1
+/// SECP256R1
+/// X25519 // Enabled with the `x25519` feature
+/// ```
+pub const ALL_KX_GROUPS: &[&dyn SupportedKxGroup] = &[
+    SECP256R1,
+    SECP384R1,
+    #[cfg(feature = "x25519")]
+    X25519,
+];
+
+// Since the type trait size cannot be determined at compile time, we must use trait objects, hence the `&dyn SupportedKxGroup`
+// annotation. Similarly, `KxGroup` must then also be taken as a reference.
+#[cfg(feature = "x25519")]
+pub const X25519: &dyn SupportedKxGroup = &KxGroup {
+    name: NamedGroup::X25519,
+    curve_type: CurveType::Curve25519,
+};
+
+pub const SECP256R1: &dyn SupportedKxGroup = &KxGroup {
+    name: NamedGroup::secp256r1,
+    curve_type: CurveType::NistP256,
+};
+
+pub const SECP384R1: &dyn SupportedKxGroup = &KxGroup {
+    name: NamedGroup::secp384r1,
+    curve_type: CurveType::NistP384,
+};
+
+/// Impl for the trait SupportedKxGroup
+///
+/// `start()` creates a new `symcrypt_sys::EcDh` struct and subsequently, a new [`KeyExchange`] struct.
+///
+/// `name()` returns the [`NamedGroup`] of the current [`KeyExchange`] group.
+impl SupportedKxGroup for KxGroup {
+    fn start(&self) -> Result<Box<(dyn ActiveKeyExchange)>, Error> {
+        let ecdh_state = EcDh::new(self.curve_type).map_err(|_| GetRandomFailed)?;
+        let mut pub_key = ecdh_state
+            .get_public_key_bytes()
+            .map_err(|_| GetRandomFailed)?;
+
+        if ecdh_state.get_curve() == CurveType::NistP256
+            || ecdh_state.get_curve() == CurveType::NistP384
```

Comment:
Maybe a match to ensure we cover all CurveTypes. Want to ensure this would be updated for a new CurveType.

Reply:
Changed